src/interfaces/token.spec.ts

```diff
@@ -0,0 +1,475 @@
+import { Contract } from "@ethersproject/contracts";
+
+import { Address, Asset, ChainId, SdkError, TokenInterface, TokenMetadata } from "..";
+import { CachedFetcher } from "../cache";
+import { Context } from "../context";
+import { assetStaticVaultV2Factory, balanceFactory, tokenFactory } from "../factories";
+import { EthAddress } from "../helpers";
+import { Yearn } from "../yearn";
+
+const getPriceUsdcMock = jest.fn();
+const getPriceFromRouterMock = jest.fn(() => Promise.resolve(1));
+const zapperBalancesMock = jest.fn();
+const zapperGasMock = jest.fn();
+const zapperSupportedTokensMock = jest.fn();
+const zapperZapInApprovalStateMock = jest.fn();
+const zapperZapInApprovalTransactionMock = jest.fn();
+const zapperZapOutApprovalStateMock = jest.fn();
+const zapperZapOutApprovalTransactionMock = jest.fn();
+const assetIconMock = jest.fn();
+const assetReadyThenMock = jest.fn();
+const metaTokensMock = jest.fn();
+const vaultsBalancesMock = jest.fn();
+const ironBankBalancesMock = jest.fn();
+
+jest.mock("@ethersproject/contracts", () => ({
+  Contract: jest.fn().mockImplementation(() => ({
+    approve: jest.fn().mockResolvedValue(true)
+  }))
+}));
+
+jest.mock("../yearn", () => ({
+  Yearn: jest.fn().mockImplementation(() => ({
+    services: {
+      asset: {
+        ready: { then: assetReadyThenMock },
+        icon: assetIconMock
+      },
+      meta: {
+        tokens: metaTokensMock
+      },
+      oracle: {
+        getPriceFromRouter: getPriceFromRouterMock,
+        getPriceUsdc: getPriceUsdcMock
+      },
+      zapper: {
+        balances: zapperBalancesMock,
+        gas: zapperGasMock,
+        supportedTokens: zapperSupportedTokensMock,
+        zapInApprovalState: zapperZapInApprovalStateMock,
+        zapInApprovalTransaction: zapperZapInApprovalTransactionMock,
+        zapOutApprovalState: zapperZapOutApprovalStateMock,
+        zapOutApprovalTransaction: zapperZapOutApprovalTransactionMock
+      }
+    },
+    ironBank: { balances: ironBankBalancesMock },
+    vaults: {
+      balances: vaultsBalancesMock
+    }
+  }))
+}));
+
+jest.mock("../context", () => ({
+  Context: jest.fn().mockImplementation(() => ({
+    provider: {
+      write: {
+        getSigner: jest.fn().mockImplementation(() => ({
+          sendTransaction: jest.fn().mockResolvedValue("transaction")
+        }))
+      }
+    }
+  }))
+}));
+
+describe("TokenInterface", () => {
+  let tokenInterface: TokenInterface<1>;
+
+  let mockedYearn: Yearn<ChainId>;
+
+  beforeEach(() => {
+    mockedYearn = new (Yearn as jest.Mock<Yearn<ChainId>>)();
+    tokenInterface = new TokenInterface(mockedYearn, 1, new Context({}));
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe("price", () => {
+    it("should should get the exchange rate between two tokens", async () => {
+      expect(await tokenInterface.price("0x000", "0x001")).toEqual(1);
```

Comment:
this tests are all quite good! Nice work on them.

Just as a nit:

Just for readability and consistency, I think we should declare a result var before actually doing the expects. Normally tests have the structure of Setup -> Run what you want to test -> Assert validations. Just for readability i would change this ones to be a bit more like 
```
 it("should should get the exchange rate between two tokens", async () => {
      const priceResult = await tokenInterface.price("0x000", "0x001");
      
      expect(priceResult).toEqual(1);
      expect(getPriceFromRouterMock).toHaveBeenCalledTimes(1);
      expect(getPriceFromRouterMock).toHaveBeenCalledWith("0x000", "0x001");
    });
  });
```

Just so we can clearly read what is going on in the test

Reply:
Good idea, improved the readability with https://github.com/yearn/yearn-sdk/pull/218/commits/1e2bf08d04a5db1f5590a54d804acd4a1e082220